// atom-languageclient/build/lib/utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const atom_1 = require("atom");
const vscode_jsonrpc_1 = require("vscode-jsonrpc");
class Utils {
    /**
     * Obtain the range of the word at the given editor position.
     * Uses the non-word characters from the position's grammar scope.
     */
    static getWordAtPosition(editor, position) {
        const scopeDescriptor = editor.scopeDescriptorForBufferPosition(position);
        const nonWordCharacters = Utils.escapeRegExp(editor.getNonWordCharacters(scopeDescriptor));
        const range = Utils._getRegexpRangeAtPosition(editor.getBuffer(), position, new RegExp(`^[\t ]*$|[^\\s${nonWordCharacters}]+`, 'g'));
        if (range == null) {
            return new atom_1.Range(position, position);
        }
        return range;
    }
    static escapeRegExp(string) {
        // From atom/underscore-plus.
        return string.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
    }
    static _getRegexpRangeAtPosition(buffer, position, wordRegex) {
        const { row, column } = position;
        const rowRange = buffer.rangeForRow(row, false);
        let matchData;
        // Extract the expression from the row text.
        buffer.scanInRange(wordRegex, rowRange, (data) => {
            const { range } = data;
            if (position.isGreaterThanOrEqual(range.start) &&
                // Range endpoints are exclusive.
                position.isLessThan(range.end)) {
                matchData = data;
                data.stop();
                return;
            }
            // Stop the scan if the scanner has passed our position.
            if (range.end.column > column) {
                data.stop();
            }
        });
        return matchData == null ? null : matchData.range;
    }
    /**
     * For the given connection and cancellationTokens map, cancel the existing
     * CancellationToken for that connection then create and store a new
     * CancellationToken to be used for the current request.
     */
    static cancelAndRefreshCancellationToken(key, cancellationTokens) {
        let cancellationToken = cancellationTokens.get(key);
        if (cancellationToken !== undefined && !cancellationToken.token.isCancellationRequested) {
            cancellationToken.cancel();
        }
        cancellationToken = new vscode_jsonrpc_1.CancellationTokenSource();
        cancellationTokens.set(key, cancellationToken);
        return cancellationToken.token;
    }
    static doWithCancellationToken(key, cancellationTokens, work) {
        return __awaiter(this, void 0, void 0, function* () {
            const token = Utils.cancelAndRefreshCancellationToken(key, cancellationTokens);
            const result = yield work(token);
            cancellationTokens.delete(key);
            return result;
        });
    }
    static assertUnreachable(_) {
        return _;
    }
}
exports.default = Utils;
